### PR TITLE
Read alternative region from environment for DynamoDB connection

### PR DESCRIPTION
**Pull request by exalted**
Closes https://github.com/architect/architect/issues/887

### We really appreciate your commitment to improving Architect

To maintain a high standard of quality in our releases, before merging every pull request we ask that you've completed the following:

- [x] Forked the repo and created your branch from `master`
- [x] Made sure tests pass (run `npm it` from the repo root)
- [ ] Expanded test coverage related to your changes:
  - [ ] Added and/or updated unit tests (if appropriate)
  - [ ] Added and/or updated integration tests (if appropriate)
- [ ] Updated relevant documentation:
  - [ ] Internal to this repo (e.g. `readme.md`, help docs, inline docs & comments, etc.)
  - [ ] [Architect docs (arc.codes)](https://github.com/architect/arc.codes)
- [ ] Summarized your changes in `changelog.md`
- [x] Linked to any related issues, PRs, etc. below that may relate to, consume, or necessitate these changes

Please also be sure to completed the CLA (if you haven't already).

Learn more about [contributing to Architect here](https://arc.codes/intro/community).

Thanks again!

### CODE REVIEW

**Comment by ryanblock** [comment]
Hey, I'm sorry I've let this sit so long. Would like to get it merged! A couple questions:

- What is the use case around this? Want to make sure we've got the right approach and assumptions here
- Any idea what implications / side-effects may occur because of this? Seems fairly straightforward (but per above, I'm not clear on the use case yet)
- Sandbox has matured to the point where I'm disinclined to accept behavior changing PRs that aren't fully surrounded in tests; what are the assumptions we're making about this change so we can figure out the best way to test for them?

Thanks!

**Comment by exalted** [comment]
Hey @ryanblock no worries, I am glad we're here at last! 🙂 

So my use case is this that I want to connect to a local DynamoDB instance (created with [DynamoDB Local](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html), specifically using the version I've installed on my Mac using this [Homebrew formula](https://github.com/Homebrew/homebrew-cask/blob/fe6ba1146f60749111750d721b86b59188ac54aa/Casks/dynamodb-local.rb)).

The reason why I sometimes prefer connecting to a local DynamoDB instance, for local development, is because I want the data to be persisted as opposed to in-memory DynamoDB which sandbox uses otherwise.

Also, I tend to use [NoSQL Workbench for DynamoDB GUI Client](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/workbench.html) to query the database (installed via Homebrew using [this formula](https://github.com/Homebrew/homebrew-cask/blob/fe6ba1146f60749111750d721b86b59188ac54aa/Casks/nosql-workbench-for-amazon-dynamodb.rb)).

So, long story short, I want to be able to;

1. Run a local (persistent) DynamoDB instance.
2. Make it so that `arc sandbox` connects to the same DynamoDB instance.
3. Make it so that NoSQL Workbench connects to the same DynamoDB instance.

The only missing piece of the puzzle is that `arc sandbox` currently hardcodes `us-west-2` as `AWS_REGION`, because I guess, it actually doesn't care much what value it has, as long as AWS CLI/SDK is "happy", however if I am not able to tweak this value, then I am not able to let `arc sandbox` to connect to the local DynamoDB instance (or any other DynamoDB instance in other regions for that matter).

I hope this is useful.

**Comment by exalted** [comment]
So yeah, I need to be able to set `AWS_REGION` to `localhost` (in my specific case) to let `arc sandbox` to connect to the local DynamoDB instance.

**Comment by ryanblock** [comment]
All that makes sense! This may be a dumb question but tbh I haven't used DDB Local, so why does it respect or require a region setting if it's running locally? And if so, why not just set it to us-west-2 since, again, it's running locally?

**Comment by exalted** [comment]
> All that makes sense! This may be a dumb question but tbh I haven't used DDB Local, so why does it respect or require a region setting if it's running locally? And if so, why not just set it to us-west-2 since, again, it's running locally?

All fair questions, and that's exactly what I tried at first, however;

1. Local database's (server) endpoint is `http://localhost:<port>`, and as far as I know there's no way to change that (so it doesn't "require" a region, instead think of its region as `localhost`, i.e., serves from `localhost` other valid `AWS_REGION` values.)
2. If you want to connect to a local database with NoSQL Workbench GUI (client), then the hostname is set to `localhost` without possibly to change it (which kinda makes sense actually).

**Comment by exalted** [comment]
I mean, in all honesty, if you reeeealy want to make something nicer here, I might accept a `arc sandbox --local-dynamodb` or something like that 🙂

**Comment by ryanblock** [comment]
I'm down with that, but I'm still unclear what's going on under the hood with the local DB. And these docs tell me absolutely nothing useful about client region with DynamoDB Local: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.UsageNotes.html

Just to level-set, the operation in question at `src/tables/_get-db-client.js` is necessary so we can pre-populate the local DynamoDB instance with the tables you've defined. We can't get a valid DynamoDB client without passing an endpoint (`http://localhost:${tablesPort}`) and region (which can be basically any string for Dynalite).

My concern is that by using an env var in here, we could potentially introduce side effects we don't fully understand (and this PR doesn't have any tests, or a clear way to test this stuff).

What is the value you would expect to use with `AWS_REGION` in this case?

Alternately, what would you expect `arc sandbox --local-dynamodb` to do inside Sandbox?

**Comment by exalted** [comment]
I am not sure what to tell you @ryanblock , what I am trying to achieve here is that IMO `arc sandbox` should care whether http://localhost:${tablesPort} is a Dynalite server or a dynamodb-local server, considering they should both adhere to same/similar APIs, for the most part…

FWIW, in order to be able to connect `arc sandbox` to a dynamodb-local server however, I need to explicit tell it to not create a Dynalite server, via `ARC_DB_EXTERNAL=true` (same if you had dynalite running via `dynalite --port 5000` rather than via initialized via code).

However a key difference between `dynalite --port 5000` (in-memory by default) and `dynamodb-local -inMemory -port 5000` is that when `arc sandbox` is run without `AWS_REGION` set to `localhost` arc won't be able to create tables structure in case of dynamodb-local, whilst with Dynalite in-memory (external) server, this doesn't seems to be an issue.

Try:

```
# Sandbox will generate tables, no issues there.
dynalite --port 5000
export ARC_DB_EXTERNAL=true
arc sandbox
```

```
# Although console output will say it did, sandbox won't actually generate tables.
dynamodb-local -inMemory -port 5000
export ARC_DB_EXTERNAL=true
arc sandbox
```

Only way I was able to make `arc sandbox` "work" with `dynamodb-local` is like so:


```
export ARC_DB_EXTERNAL=true
dynamodb-local -inMemory -port 5000
export AWS_REGION=localhost
# Patch _get-db-client.js
arc sandbox
```
